pkg/networkservice/common/dial/client.go

```diff
@@ -0,0 +1,139 @@
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dial will dial up a grpc.ClientConnInterface if a client *url.URL is provided in the ctx, retrievable by
+// clienturlctx.ClientURL(ctx) and put the resulting grpc.ClientConnInterface into the ctx using clientconn.Store(..)
+// where it can be retrieved by other chain elements using clientconn.Load(...)
+package dial
+
+import (
+	"context"
+	"time"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clientconn"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/clienturlctx"
+	"github.com/networkservicemesh/sdk/pkg/tools/clock"
+	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
+)
+
+type dialClient struct {
+	chainCtx    context.Context
+	dialOptions []grpc.DialOption
+	dialTimeout time.Duration
+}
+
+// NewClient - returns new dial chain element
+func NewClient(chainCtx context.Context, opts ...Option) networkservice.NetworkServiceClient {
+	o := &option{}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return &dialClient{
+		chainCtx:    chainCtx,
+		dialOptions: o.dialOptions,
+		dialTimeout: o.dialTimeout,
+	}
+}
+
+func (d *dialClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	// If no clientURL, we have no work to do
+	// call the next in the chain
+	clientURL := clienturlctx.ClientURL(ctx)
+	if clientURL == nil {
+		return next.Client(ctx).Request(ctx, request, opts...)
+	}
+
+	di, err := d.dial(ctx, func(di *dialInfo) {
+		_, _ = d.Close(clienturlctx.WithClientURL(ctx, di.url), request.GetConnection(), opts...)
+	})
+	if di != nil {
+		store(ctx, di)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		_ = di.Close()
+		return nil, err
+	}
+	return conn, nil
+}
+
+func (d *dialClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*emptypb.Empty, error) {
+	// If no clientURL, we have no work to do
+	clientURL := clienturlctx.ClientURL(ctx)
+	if clientURL == nil {
+		return next.Client(ctx).Close(ctx, conn, opts...)
+	}
+
+	di, _ := d.dial(ctx, func(_ *dialInfo) {})
+
+	_, err := next.Client(ctx).Close(ctx, conn, opts...)
+	if di != nil {
+		_ = di.Close()
+	}
+	clientconn.Delete(ctx)
```

Comment:
Do we need this explicit delete? Metadata for this connection would be cleaned up after this Close.

Reply:
I'm a bit paranoid about having the thing that is responsible clean up after itself.